ArjunaCore/arjuna/tests/classes/com/hp/mwtests/ts/arjuna/recovery/RecoverySuspendTest.java

```diff
@@ -0,0 +1,164 @@
+
+import java.util.Arrays;
+
+@WithByteman
+@BMUnitConfig(debug=true)
+@BMScript("recoverySuspend")
+public class RecoverySuspendTest {
+
+    private static RecoveryManager _manager;
+    private static RecordTypeMap _recordTypeMap;
+
+    // fields read with byteman
+    private final static int periodicRecoveryPeriod = 5;
+    private final static int recoveryBackoffPeriod = 1;
+
+    @BeforeAll
+    public static void beforeClass() {
+
+        RecoveryEnvironmentBean recoveryConfig = recoveryPropertyManager.getRecoveryEnvironmentBean();
+        String[] modules = {
+                // the module to test
+                "com.arjuna.ats.internal.arjuna.recovery.AtomicActionRecoveryModule"
+        };
+
+        // Let's go quick on this
+        recoveryConfig.setRecoveryBackoffPeriod(recoveryBackoffPeriod);
+        recoveryConfig.setPeriodicRecoveryPeriod(periodicRecoveryPeriod);
+        // don't sign off until the store is empty
+        recoveryConfig.setWaitForFinalRecovery(true);
+
+        // the test set of modules
+        recoveryConfig.setRecoveryModuleClassNames(Arrays.asList(modules));
+
+        // obtain a new RecoveryManager with the above config:
+        _manager = RecoveryManager.manager(RecoveryManager.INDIRECT_MANAGEMENT);
+        // recovery can start
+        _manager.startRecoveryManagerThread();
+
+        _recordTypeMap = new RecordTypeMap() {
+            @SuppressWarnings("unchecked")
+            public Class getRecordClass ()
+            {
+                return BytemanControlledRecord.class;
+            }
+
+            public int getType ()
+            {
+                return RecordType.USER_DEF_FIRST0;
+            }
+        };
+
+        RecordTypeManager.manager().add(_recordTypeMap);
+    }
+
+    @BeforeEach
+    public void beforeEach() {
+        // In case the Recovery Manager was suspended, this will resume it
+        _manager.resume();
+    }
+
+    @AfterAll
+    public static void afterClass() {
+        RecordTypeManager.manager().remove(_recordTypeMap);
+    }
+
+    @AfterEach
+    public void afterEach() {
+        if (!TxControl.isEnabled()) {
+            // Re-activate transaction creation
+            TxControl.enable();
+        }
+    }
+
+    @Test
+    public void testSuspensionWheneThereArentTxnsToRecover() {
+
+        // Make sure that the test environment is ready
+        BytemanControlledRecord.resetCommitCallCounter();
+        BytemanControlledRecord.setGreenFlag();
+
+        createAtomicAction(new BytemanControlledRecord(true));
+
+        // The Transaction System needs to be disabled, i.e. no new transactions can be created
+        TxControl.disable();
+
+        // Makes sure that the transaction reaper completed all transactions
+        TransactionReaper.transactionReaper().waitForAllTxnsToTerminate();
+
+        // Synchronization is not needed (i.e. async = true) as isWaitForFinalRecovery() == true
+        _manager.suspend(true);
```

Comment:
Please can I ask what part of this test is verifying that synchronization was not needed during the suspend?

Reply:
`isWaitForFinalRecovery() == true` overrides synchronisation (i.e. `async == true`). I should have explained this better.

Reply:
Sorry, I was not correct in what I was asking. I am more wondering how we can verify that https://github.com/jmfinelli/narayana/pull/4/files#diff-5713de2f3997387b2db2fbe85a49cbe705fe13a5a451e669b1ecc68855f24e04R236 did not return true and require the suspend to wait. So rather than synchronization,  I am wanting to verify that looping was not needed. Sorry, my wording was not correct.

Reply:
Yeah, the test I wrote is not really verifying what you highlighted. I can use a bit of byteman magic to do what you asked for but it's not going to be elegant

Reply:
Actually, I think I put together a nice solution to carry out what you wanted @tomjenkinson

Reply:
+1 - thank you!! (please consider this resolved)

Reply:
Thanks Tom